volatility3/framework/plugins/windows/pe_symbols.py

```diff
@@ -0,0 +1,437 @@
+        Attempts to resolve the symbols in `wanted_symbols` through PDB analysis
+        """
+
+        # the PDB name of the kernel file is not consistent for an exe, for example,
+        # a `ntoskrnl.exe` can have an internal PDB name of any of the ones in the following list
+        # The code attempts to find all possible PDBs to ensure the best chance of recovery
+        if mod_name == os_module_name:
+            pdb_names = ["ntkrnlmp.pdb", "ntkrnlpa.pdb", "ntkrpamp.pdb", "ntoskrnl.pdb"]
+
+        # for non-kernel files, replace the exe, sys, or dll extension with pdb
+        else:
+            pdb_names = [mod_name[:-3] + "pdb"]
+
+        for proc_layer_name, vad_start, vad_size in vads:
+            mod_symbols = None
+
+            # loop through each PDB name (will be just one for all but the kernel)
+            for pdb_name in pdb_names:
+                try:
+                    mod_symbols = pdbutil.PDBUtility.symbol_table_from_pdb(
+                        context,
+                        interfaces.configuration.path_join(config_path, mod_name),
+                        proc_layer_name,
+                        pdb_name,
+                        vad_start,
+                        vad_size,
+                    )
+
+                    if mod_symbols:
+                        break
+
+                # this exception is expected when the PDB can't be found or downloaded
+                except exceptions.VolatilityException:
+                    continue
+
+                # this is not expected - it means pdbconv broke when parsing the PDB
+                except TypeError as e:
+                    vollog.error(
+                        f"Unable to parse PDB file for module {pdb_name} -> {e}. Please file a bug on the GitHub issue tracker."
+                    )
+
+            # cannot do anything without the symbols
+            if not mod_symbols:
+                continue
+
+            mod_module = context.module(
+                mod_symbols, layer_name=proc_layer_name, offset=vad_start
+            )
+
+            # loop through each export looking for ones of interest
+            # break early if they are alll found
+            for func in wanted_symbols[mod_name]:
+                try:
+                    address = mod_module.get_symbol(func).address
+                except exceptions.SymbolError:
+                    continue
+
+                if mod_name not in found_symbols:
+                    found_symbols[mod_name] = []
+
+                found_symbols[mod_name].append((func, vad_start + address))
+                wanted_symbols[mod_name].remove(func)
+
+            # stop processing the layers (processes) if we found all the symbols for this module
+            if not wanted_symbols[mod_name]:
+                break
+
+    @staticmethod
+    def find_function_offsets(
+        context,
+        config_path: str,
+        layer_name: str,
+        symbol_table: str,
+        module_collector,
```

Comment:
Type signature of this callable would be good please...

Reply:
If the module_collector were the last parameter, this could be curried/made a partial function very easily (see a bit later on).

Reply:
Before this goes live, and the API becomes fixed (since this is a static method, therefore considered public and part of the API), could we move reorder the parameters please?

Reply:
fixed